amethyst_network/src/simulation/timing.rs

```diff
@@ -0,0 +1,148 @@
+//! Systems and resources to have a consistent, separate simulation frame rate from the ECS
+//! frame rate.
+
+use amethyst_core::{
+    ecs::{Read, System, Write},
+    timing::Time,
+};
+use std::{ops::RangeInclusive, time::Duration};
+
+/// Default number of network simulation frames per second.
+const DEFAULT_SIM_FRAME_RATE: u32 = 30;
+
+/// This system is used exclusively to update the state of the `NetworkSimulationTime` resource.
+pub struct NetworkSimulationTimeSystem;
+
+impl<'s> System<'s> for NetworkSimulationTimeSystem {
+    type SystemData = (Write<'s, NetworkSimulationTime>, Read<'s, Time>);
+
+    fn run(&mut self, (mut sim_time, game_time): Self::SystemData) {
+        sim_time.update_elapsed(game_time.delta_time());
+        sim_time.reset_frame_lag();
+        while sim_time.elapsed_duration() > sim_time.per_frame_duration() {
+            sim_time.increment_frame_number();
+        }
+    }
+}
+
+/// Resource to track the state of the network simulation separately from the ECS frame timings
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct NetworkSimulationTime {
+    /// The current simulation frame
+    frame_number: u32,
```

Comment:
Do we want users to allow setting custom frame number in order to let them replay certain number of frames?
I believe this could be handy when using interpolation techniques for multiplayer. Not sure though if it's the right place to take this into account..

Reply:
I would like to say that this is specifically keeping track of the current time (analogous to the Time struct for frame timings). Any sort of storing prior state for rollback/replay would likely exist separate from this. Though I could be convinced otherwise.